// cohere-ai/dist/ClientV2.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CohereClientV2 = void 0;
const Client_1 = require("./api/resources/v2/client/Client");
const Client_2 = require("./Client");
// this class will require manual updates over time
class CohereClientV2 {
    constructor(_options) {
        this._options = _options;
        this.client = new Client_2.CohereClient(this._options);
        this.clientV2 = new Client_1.V2Client(this._options);
        this.chat = this.clientV2.chat.bind(this.clientV2);
        this.chatStream = this.clientV2.chatStream.bind(this.clientV2);
        this.embed = this.clientV2.embed.bind(this.clientV2);
        this.rerank = this.clientV2.rerank.bind(this.clientV2);
        this.generateStream = this.client.generateStream.bind(this.clientV2);
        this.generate = this.client.generate.bind(this.clientV2);
        this.classify = this.client.classify.bind(this.clientV2);
        this.summarize = this.client.summarize.bind(this.clientV2);
        this.tokenize = this.client.tokenize.bind(this.clientV2);
        this.detokenize = this.client.detokenize.bind(this.clientV2);
        this.checkApiKey = this.client.checkApiKey.bind(this.clientV2);
        this.embedJobs = this.client.embedJobs;
        this.datasets = this.client.datasets;
        this.connectors = this.client.connectors;
        this.models = this.client.models;
        this.finetuning = this.client.finetuning;
        this.batches = this.client.batches;
        this.audio = this.client.audio;
    }
    fetch(input, init, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.client.fetch(input, init, requestOptions);
        });
    }
}
exports.CohereClientV2 = CohereClientV2;
