// langsmith/dist/_openapi_client/resources/sessions.cjs
"use strict";
// @ts-nocheck
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sessions = void 0;
const resource_js_1 = require("../core/resource.cjs");
const pagination_js_1 = require("../core/pagination.cjs");
const headers_js_1 = require("../internal/headers.cjs");
const path_js_1 = require("../internal/utils/path.cjs");
class Sessions extends resource_js_1.APIResource {
    /**
     * Create a new project.
     */
    create(params, options) {
        const { upsert, ...body } = params;
        return this._client.post('/api/v1/sessions', { query: { upsert }, body, ...options });
    }
    /**
     * Get a specific project.
     */
    retrieve(project_id, params = {}, options) {
        const { accept, ...query } = params ?? {};
        return this._client.get((0, path_js_1.path) `/api/v1/sessions/${project_id}`, {
            query,
            ...options,
            headers: (0, headers_js_1.buildHeaders)([{ ...(accept != null ? { accept: accept } : undefined) }, options?.headers]),
        });
    }
    /**
     * Update a project.
     */
    update(project_id, body, options) {
        return this._client.patch((0, path_js_1.path) `/api/v1/sessions/${project_id}`, { body, ...options });
    }
    /**
     * List all projects.
     */
    list(params = {}, options) {
        const { accept, ...query } = params ?? {};
        return this._client.getAPIList('/api/v1/sessions', (pagination_js_1.OffsetPaginationTopLevelArray), {
            query,
            ...options,
            headers: (0, headers_js_1.buildHeaders)([{ ...(accept != null ? { accept: accept } : undefined) }, options?.headers]),
        });
    }
    /**
     * Delete a specific project.
     */
    delete(project_id, options) {
        return this._client.delete((0, path_js_1.path) `/api/v1/sessions/${project_id}`, options);
    }
}
exports.Sessions = Sessions;
