// langsmith/dist/_openapi_client/resources/sessions.js
// @ts-nocheck
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from '../core/resource.js';
import { OffsetPaginationTopLevelArray, } from '../core/pagination.js';
import { buildHeaders } from '../internal/headers.js';
import { path } from '../internal/utils/path.js';
export class Sessions extends APIResource {
    /**
     * Create a new project.
     */
    create(params, options) {
        const { upsert, ...body } = params;
        return this._client.post('/api/v1/sessions', { query: { upsert }, body, ...options });
    }
    /**
     * Get a specific project.
     */
    retrieve(project_id, params = {}, options) {
        const { accept, ...query } = params ?? {};
        return this._client.get(path `/api/v1/sessions/${project_id}`, {
            query,
            ...options,
            headers: buildHeaders([{ ...(accept != null ? { accept: accept } : undefined) }, options?.headers]),
        });
    }
    /**
     * Update a project.
     */
    update(project_id, body, options) {
        return this._client.patch(path `/api/v1/sessions/${project_id}`, { body, ...options });
    }
    /**
     * List all projects.
     */
    list(params = {}, options) {
        const { accept, ...query } = params ?? {};
        return this._client.getAPIList('/api/v1/sessions', (OffsetPaginationTopLevelArray), {
            query,
            ...options,
            headers: buildHeaders([{ ...(accept != null ? { accept: accept } : undefined) }, options?.headers]),
        });
    }
    /**
     * Delete a specific project.
     */
    delete(project_id, options) {
        return this._client.delete(path `/api/v1/sessions/${project_id}`, options);
    }
}
