// @langchain/google/dist/utils/handler.cjs
let _langchain_core_callbacks_base = require("@langchain/core/callbacks/base");
//#region src/utils/handler.ts
var GoogleRequestCallbackHandler = class extends _langchain_core_callbacks_base.BaseCallbackHandler {
	customEventInfo(eventName) {
		const names = eventName.split("-");
		return {
			subEvent: names[1],
			module: names[2]
		};
	}
	handleCustomEvent(eventName, data, runId, tags, metadata) {
		if (!eventName) return;
		const eventInfo = this.customEventInfo(eventName);
		switch (eventInfo.subEvent) {
			case "request": return this.handleCustomRequestEvent(eventName, eventInfo, data, runId, tags, metadata);
			case "response": return this.handleCustomResponseEvent(eventName, eventInfo, data, runId, tags, metadata);
			case "chunk": return this.handleCustomChunkEvent(eventName, eventInfo, data, runId, tags, metadata);
			default: console.error(`Unexpected eventInfo for ${eventName} ${JSON.stringify(eventInfo, null, 1)}`);
		}
	}
};
var GoogleRequestLogger = class extends GoogleRequestCallbackHandler {
	name = "GoogleRequestLogger";
	shortenStringLength = 40;
	log(eventName, data, tags) {
		const splitLen = this.shortenStringLength;
		const half = splitLen / 2;
		const replacer = (_key, value) => {
			if (typeof value === "string" && splitLen > 0 && value.length > splitLen) return `${value.substring(0, half)}...${value.substring(value.length - half)}`;
			return value;
		};
		const tagStr = tags ? `[${tags}]` : "[]";
		console.log(`${eventName} ${tagStr} ${JSON.stringify(data, replacer, 1)}`);
	}
	handleCustomRequestEvent(eventName, _eventInfo, data, _runId, tags, _metadata) {
		this.log(eventName, data, tags);
	}
	handleCustomResponseEvent(eventName, _eventInfo, data, _runId, tags, _metadata) {
		this.log(eventName, data, tags);
	}
	handleCustomChunkEvent(eventName, _eventInfo, data, _runId, tags, _metadata) {
		this.log(eventName, data, tags);
	}
};
var GoogleRequestRecorder = class extends GoogleRequestCallbackHandler {
	name = "GoogleRequestRecorder";
	request = {};
	response = {};
	chunk = [];
	handleCustomRequestEvent(_eventName, _eventInfo, data, _runId, _tags, _metadata) {
		this.request = data;
	}
	handleCustomResponseEvent(_eventName, _eventInfo, data, _runId, _tags, _metadata) {
		this.response = data;
	}
	handleCustomChunkEvent(_eventName, _eventInfo, data, _runId, _tags, _metadata) {
		this.chunk.push(data);
	}
};
//#endregion
exports.GoogleRequestCallbackHandler = GoogleRequestCallbackHandler;
exports.GoogleRequestLogger = GoogleRequestLogger;
exports.GoogleRequestRecorder = GoogleRequestRecorder;
