// @langchain/google/dist/utils/stream_events.js
import { finalizeContentBlock } from "@langchain/core/language_models/compat";
//#region src/utils/stream_events.ts
/**
* Converts Gemini stream responses into LangChain ChatModelStreamEvents.
*
* @module
*/
async function* convertGoogleGeminiStream(source, options = {}) {
	const shouldStreamUsage = options.streamUsage ?? true;
	const blockAccumulators = /* @__PURE__ */ new Map();
	const blockKeyToIndex = /* @__PURE__ */ new Map();
	let nextBlockIndex = 0;
	let messageStarted = false;
	let usageSnapshot;
	let finishReason = "stop";
	const getOrCreateBlockIndex = (key, initial) => {
		const existing = blockKeyToIndex.get(key);
		if (existing !== void 0) return {
			index: existing,
			isNew: false
		};
		const index = nextBlockIndex++;
		blockKeyToIndex.set(key, index);
		blockAccumulators.set(index, { ...initial });
		return {
			index,
			isNew: true
		};
	};
	for await (const response of source) {
		if (!messageStarted) {
			messageStarted = true;
			yield { event: "message-start" };
		}
		const usageMetadata = response.usageMetadata ?? response.usage_metadata;
		if (shouldStreamUsage && usageMetadata) {
			const input = usageMetadata.promptTokenCount ?? 0;
			const output = usageMetadata.candidatesTokenCount ?? 0;
			usageSnapshot = {
				input_tokens: input,
				output_tokens: output,
				total_tokens: usageMetadata.totalTokenCount ?? input + output
			};
			yield {
				event: "usage",
				usage: usageSnapshot
			};
		}
		const candidate = response.candidates?.[0];
		if (candidate?.finishReason) finishReason = mapGeminiFinishReason(candidate.finishReason);
		const parts = candidate?.content?.parts;
		if (!parts) continue;
		let toolIdx = 0;
		for (const part of parts) if (part.text) if (part.thought) {
			const { index, isNew } = getOrCreateBlockIndex("reasoning", {
				type: "reasoning",
				reasoning: ""
			});
			if (isNew) yield {
				event: "content-block-start",
				index,
				content: {
					type: "reasoning",
					reasoning: ""
				}
			};
			const acc = blockAccumulators.get(index);
			acc.reasoning = (acc.reasoning ?? "") + part.text;
			yield {
				event: "content-block-delta",
				index,
				delta: {
					type: "reasoning-delta",
					reasoning: part.text
				}
			};
		} else {
			const { index, isNew } = getOrCreateBlockIndex("text", {
				type: "text",
				text: ""
			});
			if (isNew) yield {
				event: "content-block-start",
				index,
				content: {
					type: "text",
					text: ""
				}
			};
			const acc = blockAccumulators.get(index);
			acc.text = (acc.text ?? "") + part.text;
			yield {
				event: "content-block-delta",
				index,
				delta: {
					type: "text-delta",
					text: part.text
				}
			};
		}
		else if (part.functionCall) {
			const key = `tool:${toolIdx}`;
			const args = JSON.stringify(part.functionCall.args ?? {});
			const { index, isNew } = getOrCreateBlockIndex(key, {
				type: "tool_call_chunk",
				name: part.functionCall.name,
				args: "",
				index: toolIdx
			});
			if (isNew) yield {
				event: "content-block-start",
				index,
				content: {
					type: "tool_call_chunk",
					name: part.functionCall.name,
					args: "",
					index: toolIdx
				}
			};
			const acc = blockAccumulators.get(index);
			acc.args = args;
			yield {
				event: "content-block-delta",
				index,
				delta: {
					type: "block-delta",
					fields: {
						type: "tool_call_chunk",
						name: acc.name,
						args: acc.args
					}
				}
			};
			toolIdx += 1;
		}
	}
	for (const [index, acc] of blockAccumulators) yield {
		event: "content-block-finish",
		index,
		content: finalizeContentBlock(acc)
	};
	yield {
		event: "message-finish",
		reason: finishReason,
		...usageSnapshot ? { usage: usageSnapshot } : {},
		responseMetadata: { model_provider: "google" }
	};
}
function mapGeminiFinishReason(reason) {
	switch (reason.toLowerCase()) {
		case "max_tokens":
		case "max-token":
		case "max_token": return "length";
		case "safety":
		case "recitation":
		case "language":
		case "blocklist":
		case "prohibited_content":
		case "prohibited-content":
		case "spii":
		case "image_safety":
		case "image-safety":
		case "image_prohibited_content":
		case "image-prohibited-content":
		case "image_recitation":
		case "image-recitation": return "content_filter";
		default: return "stop";
	}
}
//#endregion
export { convertGoogleGeminiStream };
