// @langchain/mistralai/dist/utils/stream_events.cjs
let _langchain_core_language_models_openai_completions_stream = require("@langchain/core/language_models/openai_completions_stream");
//#region src/utils/stream_events.ts
function mistralDataToOpenAIChunk(data) {
	const choice = data.choices?.[0];
	if (!choice) return data;
	const delta = choice.delta ?? {};
	const toolCalls = delta.toolCalls ?? delta.tool_calls;
	const mappedDelta = {
		...delta,
		...toolCalls ? { tool_calls: toolCalls.map((tc, i) => ({
			index: tc.index ?? i,
			id: tc.id,
			type: "function",
			function: tc.function
		})) } : {}
	};
	return {
		id: data.id ?? "",
		object: "chat.completion.chunk",
		created: data.created ?? 0,
		model: data.model ?? "",
		choices: [{
			index: choice.index ?? 0,
			delta: mappedDelta,
			finish_reason: choice.finish_reason ?? choice.finishReason ?? null,
			logprobs: choice.logprobs ?? null
		}],
		usage: data.usage ?? null,
		system_fingerprint: data.system_fingerprint ?? null
	};
}
async function* convertMistralStream(source, options = {}) {
	async function* mapped() {
		for await (const data of source) yield mistralDataToOpenAIChunk(data);
	}
	yield* (0, _langchain_core_language_models_openai_completions_stream.convertOpenAICompletionsStream)(mapped(), {
		...options,
		provider: options.provider ?? "mistralai"
	});
}
//#endregion
exports.convertMistralStream = convertMistralStream;
