// @langchain/cohere/dist/utils/stream_events.cjs
require("../_virtual/_rolldown/runtime.cjs");
let _langchain_core_language_models_compat = require("@langchain/core/language_models/compat");
//#region src/utils/stream_events.ts
/**
* Converts Cohere chat stream events into LangChain ChatModelStreamEvents.
*
* @module
*/
async function* convertCohereStream(source, options = {}) {
	const shouldStreamUsage = options.streamUsage ?? true;
	let messageStarted = false;
	const textIndex = 0;
	let textStarted = false;
	let accumulatedText = "";
	let usageSnapshot;
	const toolBlocks = /* @__PURE__ */ new Map();
	for await (const chunk of source) {
		if (!messageStarted) {
			messageStarted = true;
			yield { event: "message-start" };
		}
		if (chunk.eventType === "text-generation" && typeof chunk.text === "string") {
			if (!textStarted) {
				textStarted = true;
				yield {
					event: "content-block-start",
					index: textIndex,
					content: {
						type: "text",
						text: ""
					}
				};
			}
			accumulatedText += chunk.text;
			yield {
				event: "content-block-delta",
				index: textIndex,
				delta: {
					type: "text-delta",
					text: chunk.text
				}
			};
		} else if (chunk.eventType === "stream-end") {
			const response = chunk.response ?? {};
			if (shouldStreamUsage && response.meta?.tokens) {
				const input = response.meta.tokens.inputTokens ?? 0;
				const output = response.meta.tokens.outputTokens ?? 0;
				usageSnapshot = {
					input_tokens: input,
					output_tokens: output,
					total_tokens: input + output
				};
				yield {
					event: "usage",
					usage: usageSnapshot
				};
			}
			const toolCalls = response.toolCalls ?? [];
			for (let i = 0; i < toolCalls.length; i++) {
				const tc = toolCalls[i];
				const index = textStarted ? i + 1 : i;
				const args = typeof tc.function?.arguments === "string" ? tc.function.arguments : JSON.stringify(tc.function?.arguments ?? {});
				const initial = {
					type: "tool_call_chunk",
					id: tc.id,
					name: tc.function?.name,
					args,
					index
				};
				toolBlocks.set(index, { ...initial });
				yield {
					event: "content-block-start",
					index,
					content: initial
				};
			}
		} else yield {
			event: "provider",
			provider: "cohere",
			name: chunk.eventType ?? "unknown",
			payload: chunk
		};
	}
	if (textStarted) yield {
		event: "content-block-finish",
		index: textIndex,
		content: {
			type: "text",
			text: accumulatedText
		}
	};
	for (const [index, acc] of toolBlocks) yield {
		event: "content-block-finish",
		index,
		content: (0, _langchain_core_language_models_compat.finalizeContentBlock)(acc)
	};
	yield {
		event: "message-finish",
		reason: "stop",
		...usageSnapshot ? { usage: usageSnapshot } : {},
		responseMetadata: { model_provider: "cohere" }
	};
}
//#endregion
exports.convertCohereStream = convertCohereStream;
