// langsmith/dist/_openapi_client/resources/sandboxes/registries.js
// @ts-nocheck
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from '../../core/resource.js';
import { buildHeaders } from '../../internal/headers.js';
import { path } from '../../internal/utils/path.js';
export class Registries extends APIResource {
    /**
     * Create a sandbox registry for pulling private images.
     */
    create(body, options) {
        return this._client.post('/v2/sandboxes/registries', { body, ...options });
    }
    /**
     * Get a sandbox registry by name.
     */
    retrieve(name, options) {
        return this._client.get(path `/v2/sandboxes/registries/${name}`, options);
    }
    /**
     * Update a sandbox registry's name and/or credentials.
     */
    update(name, body, options) {
        return this._client.patch(path `/v2/sandboxes/registries/${name}`, { body, ...options });
    }
    /**
     * List sandbox registries for pulling private images.
     */
    list(query = {}, options) {
        return this._client.get('/v2/sandboxes/registries', { query, ...options });
    }
    /**
     * Delete a sandbox registry by name.
     */
    delete(name, options) {
        return this._client.delete(path `/v2/sandboxes/registries/${name}`, {
            ...options,
            headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
        });
    }
}
