// @langchain/google/dist/chat_models/base.cjs
const require_misc = require("../utils/misc.cjs");
const require_errors = require("../utils/errors.cjs");
const require_stream_events = require("../utils/stream_events.cjs");
const require_profiles = require("./profiles.cjs");
const require_types = require("./types.cjs");
const require_stream = require("../utils/stream.cjs");
const require_messages = require("../converters/messages.cjs");
const require_tools = require("../converters/tools.cjs");
const require_params = require("../converters/params.cjs");
const require_metadata = require("../utils/metadata.cjs");
let _langchain_core_language_models_chat_models = require("@langchain/core/language_models/chat_models");
let _langchain_core_messages = require("@langchain/core/messages");
let _langchain_core_utils_stream = require("@langchain/core/utils/stream");
let _langchain_core_outputs = require("@langchain/core/outputs");
let eventsource_parser_stream = require("eventsource-parser/stream");
let _langchain_core_runnables = require("@langchain/core/runnables");
let _langchain_core_utils_json_schema = require("@langchain/core/utils/json_schema");
let _langchain_core_utils_types = require("@langchain/core/utils/types");
let _langchain_core_utils_standard_schema = require("@langchain/core/utils/standard_schema");
let _langchain_core_language_models_structured_output = require("@langchain/core/language_models/structured_output");
//#region src/chat_models/base.ts
function getPlatformType(platform, hasApiKey) {
	if (typeof platform !== "undefined") return platform;
	else if (hasApiKey) return "gai";
	else return "gcp";
}
function usageMetadataToTokenUsage(usageMetadata) {
	return {
		promptTokens: usageMetadata.input_tokens,
		completionTokens: usageMetadata.output_tokens,
		totalTokens: usageMetadata.total_tokens
	};
}
function mapDetailToMediaResolution(detail) {
	switch (detail) {
		case "low": return "MEDIA_RESOLUTION_LOW";
		case "high": return "MEDIA_RESOLUTION_HIGH";
		default: return;
	}
}
var BaseChatGoogle = class extends _langchain_core_language_models_chat_models.BaseChatModel {
	model;
	streaming;
	disableStreaming;
	streamUsage = true;
	_platform;
	_endpoint;
	_customHeaders;
	_location;
	_apiVersion;
	apiClient;
	constructor(params) {
		super(params);
		this.params = params;
		this._addVersion("@langchain/google", "0.2.1");
		if (!params.apiClient) throw new require_errors.ConfigurationError("BaseChatGoogle requires an apiClient. This should be provided automatically by ChatGoogle constructors. If you're extending BaseChatGoogle directly, please provide an apiClient instance.");
		this.apiClient = params.apiClient;
		this.model = params.model;
		this._platform = require_params.convertParamsToPlatformType(params);
		this._endpoint = params.endpoint;
		this._customHeaders = params.customHeaders;
		this._location = params.location;
		this._apiVersion = params.apiVersion;
		this.disableStreaming = params?.disableStreaming === true;
		this.streaming = params?.streaming === true;
		if (this.disableStreaming) this.streaming = false;
		if (params?.streaming === false) this.disableStreaming = true;
		this.streamUsage = params?.streamUsage ?? this.streamUsage;
		if (this.disableStreaming) this.streamUsage = false;
	}
	_llmType() {
		return "google";
	}
	get platformType() {
		return this._platform;
	}
	get platform() {
		return getPlatformType(this._platform, this.apiClient.hasApiKey());
	}
	get isVertexExpress() {
		return this.platform === "gcp" && this.apiClient.hasApiKey();
	}
	get apiVersion() {
		if (typeof this._apiVersion !== "undefined") return this._apiVersion;
		else if (this.platform === "gai") return "v1beta";
		else return "v1";
	}
	get location() {
		return this._location || "global";
	}
	get endpoint() {
		if (typeof this._endpoint !== "undefined") return this._endpoint;
		else if (this.platform === "gai") return "generativelanguage.googleapis.com";
		else if (this.isVertexExpress) return "aiplatform.googleapis.com";
		else if (this.location === "global") return "aiplatform.googleapis.com";
		else return `${this.location}-aiplatform.googleapis.com`;
	}
	get publisher() {
		return "google";
	}
	get urlMethod() {
		return this.streaming ? "streamGenerateContent?alt=sse" : "generateContent";
	}
	/**
	* Returns the model profile for this instance's model, describing its
	* capabilities and constraints (e.g. max tokens, input/output modalities,
	* tool calling support).
	*
	* @example
	* ```typescript
	* const model = new ChatGoogle({ model: "gemini-2.5-pro" });
	* const profile = model.profile;
	* console.log(profile.maxInputTokens); // 1048576
	* console.log(profile.imageInputs); // true
	* ```
	*/
	get profile() {
		return require_profiles.default[this.model] ?? {};
	}
	async buildUrlGemini(urlMethod) {
		return `https://${this.endpoint}/${this.apiVersion}/models/${this.model}:${urlMethod ?? this.urlMethod}`;
	}
	async buildUrlVertexExpress(urlMethod) {
		return `https://${this.endpoint}/${this.apiVersion}/publishers/${this.publisher}/models/${this.model}:${urlMethod ?? this.urlMethod}`;
	}
	async buildUrlVertexLocation(urlMethod) {
		const projectId = await this.apiClient.getProjectId();
		return `https://${this.endpoint}/${this.apiVersion}/projects/${projectId}/locations/${this.location}/publishers/${this.publisher}/models/${this.model}:${urlMethod ?? this.urlMethod}`;
	}
	async buildUrlVertex(urlMethod) {
		if (this.isVertexExpress) return this.buildUrlVertexExpress(urlMethod);
		else return this.buildUrlVertexLocation(urlMethod);
	}
	async buildUrl(urlMethod) {
		switch (this.platform) {
			case "gai": return this.buildUrlGemini(urlMethod);
			default: return this.buildUrlVertex(urlMethod);
		}
	}
	invocationParams(options) {
		const fields = combineGoogleChatModelFields(this.params, options);
		const mediaResolution = fields.mediaResolution ?? mapDetailToMediaResolution(fields.detail);
		const tools = fields.tools ? require_tools.convertToolsToGeminiTools(fields.tools) : void 0;
		const toolConfig = require_tools.convertToolChoiceToGeminiConfig(options.tool_choice, !!(tools && tools.length > 0));
		let responseJsonSchema;
		let responseMimeType;
		if (fields.responseSchema) {
			if ((0, _langchain_core_utils_types.isInteropZodSchema)(fields.responseSchema) || (0, _langchain_core_utils_standard_schema.isSerializableSchema)(fields.responseSchema)) responseJsonSchema = (0, _langchain_core_utils_json_schema.toJsonSchema)(fields.responseSchema);
			else responseJsonSchema = fields.responseSchema;
			responseMimeType = "application/json";
		}
		return {
			...tools && tools.length > 0 ? { tools } : {},
			...toolConfig ? { toolConfig } : {},
			safetySettings: fields.safetySettings,
			generationConfig: {
				temperature: fields.temperature,
				topP: fields.topP,
				topK: fields.topK,
				maxOutputTokens: fields.maxOutputTokens,
				presencePenalty: fields.presencePenalty,
				frequencyPenalty: fields.frequencyPenalty,
				stopSequences: fields.stopSequences,
				...responseMimeType ? { responseMimeType } : {},
				...responseJsonSchema ? { responseJsonSchema } : {},
				...fields.responseModalities && fields.responseModalities.length > 0 ? { responseModalities: fields.responseModalities } : {},
				candidateCount: 1,
				seed: fields.seed,
				responseLogprobs: fields.responseLogprobs,
				logprobs: fields.logprobs,
				...fields.enableEnhancedCivicAnswers !== void 0 ? { enableEnhancedCivicAnswers: fields.enableEnhancedCivicAnswers } : {},
				thinkingConfig: require_params.convertFieldsToThinkingConfig(this.model, fields),
				speechConfig: require_params.convertFieldsToSpeechConfig(fields),
				...fields.imageConfig ? { imageConfig: fields.imageConfig } : {},
				...mediaResolution ? { mediaResolution } : {}
			},
			...require_params.convertFieldsToServiceTier(this.platform, fields)
		};
	}
	getLsParams(options) {
		const params = this.invocationParams(options);
		return {
			ls_provider: this.platform === "gcp" ? "google_vertexai" : "google_genai",
			ls_model_name: this.model,
			ls_model_type: "chat",
			ls_temperature: params.generationConfig?.temperature ?? void 0,
			ls_max_tokens: params.generationConfig?.maxOutputTokens ?? void 0,
			ls_stop: options.stop
		};
	}
	getHeaders(options) {
		const fields = combineGoogleChatModelFields(this.params, options);
		return {
			"Content-Type": "application/json",
			...this.platform === "gcp" && typeof fields.serviceTier !== "undefined" && require_types.settableServiceTier.includes(fields.serviceTier) ? { "X-Vertex-AI-LLM-Shared-Request-Type": fields.serviceTier } : {},
			...this._customHeaders,
			...options.customHeaders
		};
	}
	async _generate(messages, options, runManager) {
		if (this.streaming) {
			const stream = await this._streamResponseChunks(messages, options, runManager);
			let finalChunk = null;
			for await (const chunk of stream) finalChunk = !finalChunk ? chunk : (0, _langchain_core_utils_stream.concat)(finalChunk, chunk);
			if (typeof finalChunk?.message?.content === "string" && typeof (finalChunk?.message?.additional_kwargs?.originalTextContentBlock)?.text === "string") finalChunk.message.additional_kwargs.originalTextContentBlock.text = finalChunk.message.content;
			return { generations: finalChunk ? [finalChunk] : [] };
		}
		const url = await this.buildUrl();
		const headers = this.getHeaders(options);
		const body = {
			...this.invocationParams(options),
			systemInstruction: require_messages.convertMessagesToGeminiSystemInstruction(messages),
			contents: require_messages.convertMessagesToGeminiContents(messages)
		};
		const moduleName = this.constructor.name;
		await runManager?.handleCustomEvent(`google-request-${moduleName}`, {
			url,
			headers,
			body
		});
		let response;
		try {
			response = await this.caller.callWithOptions({ signal: options.signal }, async () => {
				const nextResponse = await this.apiClient.fetch(new Request(url, {
					method: "POST",
					headers,
					body: JSON.stringify(body),
					signal: options.signal
				}));
				if (!nextResponse.ok) throw await require_errors.RequestError.fromResponse(nextResponse);
				return nextResponse;
			});
		} catch (error) {
			await runManager?.handleCustomEvent(`google-response-${moduleName}`, { error });
			throw error;
		}
		const data = await response.json();
		await runManager?.handleCustomEvent(`google-response-${moduleName}`, {
			data,
			url: response.url,
			headers: Array.from(response.headers.entries()),
			status: response.status,
			statusText: response.statusText
		});
		if (data.promptFeedback?.blockReason) throw require_errors.PromptBlockedError.fromPromptFeedback(data.promptFeedback);
		if (!data.candidates || data.candidates.length === 0) throw new require_errors.NoCandidatesError();
		const candidate = data.candidates[0];
		const message = require_messages.convertGeminiCandidateToAIMessage(candidate);
		const text = require_messages.convertAIMessageToText(message);
		const usageMetadata = require_messages.convertGeminiGenerateContentResponseToUsageMetadata(data);
		message.usage_metadata = usageMetadata;
		const serviceTier = require_misc.iife(() => {
			const trafficType = data.usageMetadata?.trafficType;
			const serviceTierHeader = response.headers.get("x-gemini-service-tier");
			if (trafficType?.startsWith("ON_DEMAND_")) return trafficType?.substring(10).toLowerCase();
			return serviceTierHeader || "standard";
		});
		return {
			generations: [{
				text,
				message,
				generationInfo: {
					finishReason: candidate.finishReason,
					finishMessage: candidate.finishMessage,
					safetyRatings: candidate.safetyRatings,
					citationMetadata: candidate.citationMetadata,
					tokenCount: candidate.tokenCount
				}
			}],
			llmOutput: {
				tokenUsage: usageMetadataToTokenUsage(usageMetadata),
				model: data.modelVersion,
				responseId: data.responseId,
				usageMetadata,
				serviceTier
			}
		};
	}
	async *_streamChatModelEvents(messages, options, _runManager) {
		const body = {
			...this.invocationParams(options),
			systemInstruction: require_messages.convertMessagesToGeminiSystemInstruction(messages),
			contents: require_messages.convertMessagesToGeminiContents(messages)
		};
		const url = await this.buildUrl("streamGenerateContent?alt=sse");
		const headers = this.getHeaders(options);
		const response = await this.apiClient.fetch(new Request(url, {
			method: "POST",
			headers,
			body: JSON.stringify(body),
			signal: options.signal
		}));
		if (!response.ok) throw await require_errors.RequestError.fromResponse(response);
		if (!response.body) return;
		const eventStream = response.body.pipeThrough(new TextDecoderStream()).pipeThrough(new eventsource_parser_stream.EventSourceParserStream()).pipeThrough(new require_stream.SafeJsonEventParserStream());
		const shouldStreamUsage = this.streamUsage !== false && options.streamUsage !== false;
		async function* geminiChunks(stream, signal) {
			const reader = stream.getReader();
			try {
				while (true) {
					if (signal?.aborted) return;
					const { done, value } = await reader.read();
					if (done) break;
					if (value !== null) yield value;
				}
			} finally {
				reader.releaseLock();
			}
		}
		yield* require_stream_events.convertGoogleGeminiStream(geminiChunks(eventStream, options.signal), { streamUsage: shouldStreamUsage });
	}
	async *_streamResponseChunks(messages, options, runManager) {
		const streamUsage = this.streamUsage ?? true;
		const body = {
			...this.invocationParams(options),
			systemInstruction: require_messages.convertMessagesToGeminiSystemInstruction(messages),
			contents: require_messages.convertMessagesToGeminiContents(messages)
		};
		const url = await this.buildUrl("streamGenerateContent?alt=sse");
		const headers = this.getHeaders(options);
		const moduleName = this.constructor.name;
		await runManager?.handleCustomEvent(`google-request-${moduleName}`, {
			url,
			headers,
			body
		});
		let response;
		try {
			response = await this.caller.callWithOptions({ signal: options.signal }, async () => {
				const nextResponse = await this.apiClient.fetch(new Request(url, {
					method: "POST",
					headers,
					body: JSON.stringify(body),
					signal: options.signal
				}));
				if (!nextResponse.ok) throw await require_errors.RequestError.fromResponse(nextResponse);
				return nextResponse;
			});
		} catch (error) {
			await runManager?.handleCustomEvent(`google-response-${moduleName}`, { error });
			throw error;
		}
		await runManager?.handleCustomEvent(`google-response-${moduleName}`, {
			url: response.url,
			headers: response.headers,
			status: response.status,
			statusText: response.statusText
		});
		if (response.body) {
			let previousUsage;
			const reader = response.body.pipeThrough(new TextDecoderStream()).pipeThrough(new eventsource_parser_stream.EventSourceParserStream()).pipeThrough(new require_stream.SafeJsonEventParserStream()).pipeThrough(new TransformStream({ transform(chunk, controller) {
				runManager?.handleCustomEvent(`google-chunk-${moduleName}`, { chunk });
				if (chunk === null) {
					controller.enqueue(new _langchain_core_outputs.ChatGenerationChunk({
						text: "",
						message: new _langchain_core_messages.AIMessageChunk({ content: "" }),
						generationInfo: { finishReason: "stop" }
					}));
					return;
				}
				const candidate = chunk.candidates?.[0];
				if (!candidate) return;
				const message = require_messages.convertGeminiCandidateToAIMessage(candidate);
				const text = require_messages.convertAIMessageToText(message);
				const parts = candidate.content?.parts ?? [];
				const toolCalls = require_messages.convertGeminiPartsToToolCalls(parts);
				if (parts.length > 0 || candidate.finishReason) {
					const messageChunkParams = {
						content: message.content,
						tool_calls: toolCalls,
						response_metadata: { model_provider: "google" },
						additional_kwargs: {
							...message.additional_kwargs.originalTextContentBlock ? { originalTextContentBlock: message.additional_kwargs.originalTextContentBlock } : {},
							...candidate.finishReason ? {
								finishReason: candidate.finishReason,
								finishMessage: candidate.finishMessage
							} : {}
						}
					};
					if (chunk?.usageMetadata && streamUsage) {
						const cumulative = require_messages.convertGeminiGenerateContentResponseToUsageMetadata(chunk);
						messageChunkParams.usage_metadata = require_metadata.subtractUsageMetadata(cumulative, previousUsage);
						previousUsage = cumulative;
					}
					const messageChunk = new _langchain_core_messages.AIMessageChunk(messageChunkParams);
					controller.enqueue(new _langchain_core_outputs.ChatGenerationChunk({
						text,
						message: messageChunk,
						generationInfo: {
							...candidate.finishReason && {
								finishReason: candidate.finishReason,
								finishMessage: candidate.finishMessage
							},
							...candidate.safetyRatings && { safetyRatings: candidate.safetyRatings }
						}
					}));
				}
			} })).getReader();
			try {
				while (true) {
					const { done, value } = await reader.read();
					if (done) break;
					yield value;
					await runManager?.handleLLMNewToken(value.text ?? "", void 0, void 0, void 0, void 0, { chunk: value });
				}
			} finally {
				reader.releaseLock();
			}
		}
	}
	/**
	* Bind tool-like objects to this chat model.
	*
	* @param tools A list of tool definitions to bind to this chat model.
	* Can be a structured tool, an OpenAI formatted tool, or a Gemini function declaration.
	* @param kwargs Any additional parameters to bind.
	*/
	bindTools(tools, kwargs) {
		return this.withConfig({
			...kwargs,
			tools
		});
	}
	/**
	* Get structured output from the model based on a schema.
	*
	* This method supports two modes:
	* - `jsonMode`: Uses `responseSchema` to get JSON responses directly (preferred for structured outputs)
	* - `functionCalling`: Uses function calling with tools (default, for compatibility)
	*
	* @param outputSchema - The schema to use for structured output (Zod schema or JSON Schema)
	* @param config - Configuration options including method and whether to include raw response
	* @returns A Runnable that returns the parsed structured output
	*/
	withStructuredOutput(outputSchema, config) {
		let llm;
		let outputParser;
		const { schema, name, includeRaw } = {
			...config,
			schema: outputSchema
		};
		let method = config?.method ?? "functionCalling";
		if (method === "jsonMode") {
			console.warn(`"jsonMode" is not supported for Google models. Falling back to "jsonSchema".`);
			method = "jsonSchema";
		}
		if (method === "jsonSchema") {
			llm = this.withConfig({ responseSchema: schema });
			outputParser = _langchain_core_runnables.RunnableLambda.from(async (input) => {
				if (!_langchain_core_messages.AIMessage.isInstance(input) && !_langchain_core_messages.AIMessageChunk.isInstance(input)) throw new require_errors.MalformedOutputError({ message: "Input is not an AIMessage or AIMessageChunk." });
				if (!input.text) throw new require_errors.MalformedOutputError({ message: "No content found in response. Cannot parse structured output." });
				return await (0, _langchain_core_language_models_structured_output.createContentParser)(schema).parse(input.text);
			});
		} else if (method === "functionCalling") {
			let functionName = name ?? "extract";
			let geminiFunctionDeclaration;
			if ((0, _langchain_core_utils_types.isInteropZodSchema)(schema) || (0, _langchain_core_utils_standard_schema.isSerializableSchema)(schema)) {
				const jsonSchema = require_tools.schemaToGeminiParameters(schema);
				const description = typeof jsonSchema.description === "string" ? jsonSchema.description : "A function available to call.";
				geminiFunctionDeclaration = {
					name: functionName,
					description,
					parameters: jsonSchema
				};
			} else if (typeof schema.name === "string" && typeof schema.parameters === "object" && schema.parameters != null) {
				geminiFunctionDeclaration = schema;
				functionName = schema.name;
			} else {
				const parameters = require_tools.schemaToGeminiParameters(schema);
				geminiFunctionDeclaration = {
					name: functionName,
					description: schema.description ?? "",
					parameters
				};
			}
			const tools = [{ functionDeclarations: [geminiFunctionDeclaration] }];
			llm = this.bindTools(tools).withConfig({ tool_choice: functionName });
			outputParser = (0, _langchain_core_language_models_structured_output.createFunctionCallingParser)(schema, functionName);
		} else throw new require_errors.ConfigurationError(`Unrecognized structured output method '${method}'. Expected 'functionCalling' or 'jsonSchema'`);
		return (0, _langchain_core_language_models_structured_output.assembleStructuredOutputPipeline)(llm, outputParser, includeRaw, includeRaw ? "ChatGoogleStructuredOutputRunnable" : "ChatGoogleStructuredOutput");
	}
};
function combineGoogleChatModelFields(a, b, ...rest) {
	const combined = {
		temperature: b.temperature ?? a.temperature,
		topP: b.topP ?? a.topP,
		topK: b.topK ?? a.topK,
		maxOutputTokens: b.maxOutputTokens ?? a.maxOutputTokens,
		presencePenalty: b.presencePenalty ?? a.presencePenalty,
		frequencyPenalty: b.frequencyPenalty ?? a.frequencyPenalty,
		stopSequences: b.stopSequences ?? a.stopSequences,
		seed: b.seed ?? a.seed,
		responseLogprobs: b.responseLogprobs ?? a.responseLogprobs,
		logprobs: b.logprobs ?? a.logprobs,
		safetySettings: b.safetySettings ?? a.safetySettings,
		responseSchema: b.responseSchema ?? a.responseSchema,
		tools: b.tools ?? a.tools,
		responseModalities: b.responseModalities ?? a.responseModalities,
		enableEnhancedCivicAnswers: b.enableEnhancedCivicAnswers ?? a.enableEnhancedCivicAnswers,
		speechConfig: b.speechConfig ?? a.speechConfig,
		imageConfig: b.imageConfig ?? a.imageConfig,
		mediaResolution: b.mediaResolution ?? a.mediaResolution,
		detail: b.detail ?? a.detail,
		maxReasoningTokens: b.maxReasoningTokens ?? a.maxReasoningTokens,
		thinkingBudget: b.thinkingBudget ?? a.thinkingBudget,
		reasoningEffort: b.reasoningEffort ?? a.reasoningEffort,
		thinkingLevel: b.thinkingLevel ?? a.thinkingLevel,
		serviceTier: b.serviceTier ?? a.serviceTier
	};
	if (rest.length > 0) return combineGoogleChatModelFields(combined, rest[0], ...rest.slice(1));
	return combined;
}
function getGoogleChatModelParams(modelOrParams, paramsArg) {
	return {
		model: typeof modelOrParams === "string" ? modelOrParams : modelOrParams.model,
		...typeof modelOrParams === "string" ? paramsArg : modelOrParams
	};
}
//#endregion
exports.BaseChatGoogle = BaseChatGoogle;
exports.getGoogleChatModelParams = getGoogleChatModelParams;
exports.getPlatformType = getPlatformType;
