// langsmith/dist/_openapi_client/resources/sandboxes/registries.cjs
"use strict";
// @ts-nocheck
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Registries = void 0;
const resource_js_1 = require("../../core/resource.cjs");
const headers_js_1 = require("../../internal/headers.cjs");
const path_js_1 = require("../../internal/utils/path.cjs");
class Registries extends resource_js_1.APIResource {
    /**
     * Create a sandbox registry for pulling private images.
     */
    create(body, options) {
        return this._client.post('/v2/sandboxes/registries', { body, ...options });
    }
    /**
     * Get a sandbox registry by name.
     */
    retrieve(name, options) {
        return this._client.get((0, path_js_1.path) `/v2/sandboxes/registries/${name}`, options);
    }
    /**
     * Update a sandbox registry's name and/or credentials.
     */
    update(name, body, options) {
        return this._client.patch((0, path_js_1.path) `/v2/sandboxes/registries/${name}`, { body, ...options });
    }
    /**
     * List sandbox registries for pulling private images.
     */
    list(query = {}, options) {
        return this._client.get('/v2/sandboxes/registries', { query, ...options });
    }
    /**
     * Delete a sandbox registry by name.
     */
    delete(name, options) {
        return this._client.delete((0, path_js_1.path) `/v2/sandboxes/registries/${name}`, {
            ...options,
            headers: (0, headers_js_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
}
exports.Registries = Registries;
