// @langchain/google/dist/converters/params.js
import { settableServiceTier } from "../chat_models/types.js";
//#region src/converters/params.ts
/**
* Converts a reasoning "effort" level value (e.g., "minimal", "low", "medium", "high")
* to a corresponding reasoning token budget, depending on the target Gemini model.
*
* This mapping is essential when configuring Google Gemini API calls with reasoning
* effort/level fields (as in the Gemini OpenAI-compatible API) or when translating
* user-friendly configuration to explicit token budgets.
*
* - The mapping is refined to reflect the constraints documented by Google:
*   - For "gemini-2.5-flash" models, token budgets may be set as low as 0 and up to 24k.
*   - For other Gemini models, minimum is 128, maximum is 32k.
* - Standard effort levels and their token mappings:
*   - "none" / "minimal": lowest allowed (minTokens)
*   - "low": 1k tokens
*   - "medium": 8k tokens
*   - "high": maximum allowed for the model (maxTokens)
* - See reference:
*   https://ai.google.dev/gemini-api/docs/thinking#thinking-levels
*   https://ai.google.dev/gemini-api/docs/openai#thinking
*
* @param {string | undefined} modelName
*   The Gemini model name (e.g., "gemini-2.5-flash"). Used to determine token range constraints.
* @param {string | undefined} effort
*   The reasoning effort level: one of "none", "minimal", "low", "medium", "high"
*   (case-insensitive). If undefined, the function returns undefined.
* @returns {number | undefined}
*   The corresponding token count for the given effort and model. Returns undefined if
*   the effort is not recognized or not provided.
*
* @example
* convertReasoningEffortToReasoningTokens("gemini-2.5-flash", "medium") // 8192
* convertReasoningEffortToReasoningTokens("gemini-1.5-pro", "high") // 32768
* convertReasoningEffortToReasoningTokens("gemini-2.5-flash", "none") // 0
* convertReasoningEffortToReasoningTokens(undefined, "low") // 1024
*/
function convertReasoningEffortToReasoningTokens(modelName, effort) {
	if (effort === void 0) return;
	const minTokens = modelName?.startsWith("gemini-2.5-flash") ? 0 : 128;
	const maxTokens = modelName?.startsWith("gemini-2.5-flash") ? 24 * 1024 : 32 * 1024;
	switch (effort.toLowerCase()) {
		case "none":
		case "minimal": return minTokens;
		case "low": return 1024;
		case "medium": return 8 * 1024;
		case "high": return maxTokens;
		default: return;
	}
}
/**
* Converts a reasoning token budget (e.g., 0, 1024, 8192, 32768) to its corresponding
* reasoning "effort" level, as used in the Google Gemini API ("minimal", "low", "medium", "high").
*
* This is the inverse of {@link convertReasoningEffortToReasoningTokens}. It is useful for
* translating explicit numeric settings (often surfaced in raw API responses or configs)
* into user-facing friendly reasoning levels/labels. The mapping relies on Google's
* documented reasoning levels for different models:
*
* - -1 means default effort (per API documentation: this is "minimal" for flash-lite models
*   and "high" (dynamic) for others).
* - 0 means "minimal" unless the model is "gemini-3-pro", for which it maps to "low".
* - <= 1024 tokens: "low"
* - <= 8192 tokens: "medium" (unless model is "gemini-3-pro", for which it reverts to "low")
* - > 8192: "high"
*
* Reference:
*   https://ai.google.dev/gemini-api/docs/thinking#thinking-levels
*   https://ai.google.dev/gemini-api/docs/openai#thinking
*
* @param {string | undefined} model
*   The Gemini model name (e.g., "gemini-3-pro"). Used for model-specific special cases.
* @param {number | undefined} reasoningTokens
*   The token budget to convert. If undefined, the function returns undefined.
* @returns {Lowercase<Gemini.ThinkingLevel> | undefined}
*   The reasoning "effort" level ("minimal", "low", "medium", "high") or undefined if input is not valid.
*
* @example
* convertReasoningTokensToReasoningEffort("gemini-3-pro", 0) // "low"
* convertReasoningTokensToReasoningEffort("gemini-1.5-pro", 0) // "minimal"
* convertReasoningTokensToReasoningEffort(undefined, 1024) // "low"
* convertReasoningTokensToReasoningEffort(undefined, -1) // "high"
* convertReasoningTokensToReasoningEffort(undefined, 9000) // "high"
*/
function convertReasoningTokensToReasoningEffort(model, reasoningTokens) {
	if (typeof reasoningTokens === "undefined") return;
	else if (reasoningTokens === -1) {
		if (model && getModelLevel(model) === "flash-lite") return "minimal";
		return "high";
	} else if (reasoningTokens === 0) if (model?.startsWith("gemini-3-pro")) return "low";
	else return "minimal";
	else if (reasoningTokens <= 1024) return "low";
	else if (reasoningTokens <= 8192) if (model?.startsWith("gemini-3-pro")) return "low";
	else return "medium";
	else return "high";
}
/**
* Infers and returns the Google platform type from the provided chat model parameters.
*
* This function helps determine which Google platform (e.g., Vertex AI on GCP or the public API)
* should be used for calling Gemini/PaLM endpoints, based on the options provided in
* {@link BaseChatGoogleParams} (such as model constructors or call options).
*/
function convertParamsToPlatformType(params) {
	if (typeof params === "undefined") return;
	if (typeof params.platformType !== "undefined") return params.platformType;
	if (params.vertexai === true) return "gcp";
}
/**
* Returns the model version from a Gemini model name.
* e.g. "gemini-2.5-flash" → "2.5"
*/
function getModelVersion(model) {
	return model.split("-")[1];
}
/**
* Returns the model level from a Gemini model name.
* e.g. "gemini-2.5-flash" → "flash", "gemini-2.5-flash-lite" → "flash-lite"
*/
function getModelLevel(model) {
	const parts = model.split("-");
	let ret = parts[2];
	if (ret === "flash" && parts[3] === "lite") ret = "flash-lite";
	return ret;
}
/**
* Returns the model specialty from a Gemini model name.
* e.g. "gemini-2.5-flash-preview-image" → "image",
*      "gemini-2.5-flash-preview-tts" → "tts",
*      "gemini-2.5-flash" → ""
*/
function getModelSpecialty(model) {
	if (model.includes("image")) return "image";
	else if (model.includes("tts")) return "tts";
	else return "";
}
/**
* Builds the `thinkingConfig` object for the Gemini `generationConfig`,
* translating user-facing reasoning fields into the wire format.
*
* @param model  The full Gemini model name (e.g. "gemini-2.5-flash").
* @param fields The combined user/call-option fields.
* @returns A populated {@link Gemini.ThinkingConfig}, or `{}`
*          when thinking should not be explicitly configured.
*/
function convertFieldsToThinkingConfig(model, fields) {
	let includeThoughts = true;
	let thinkingBudget = fields?.maxReasoningTokens ?? fields?.thinkingBudget ?? convertReasoningEffortToReasoningTokens(model, fields?.reasoningEffort ?? fields?.thinkingLevel);
	if (thinkingBudget === 0 || model.includes("pro") && thinkingBudget === 128) includeThoughts = false;
	if (model.startsWith("gemini-2.5-pro") && typeof thinkingBudget !== "undefined") {
		if (thinkingBudget >= 0 && thinkingBudget < 128) thinkingBudget = 128;
	}
	let thinkingLevel = (fields?.reasoningEffort ?? fields?.thinkingLevel ?? convertReasoningTokensToReasoningEffort(model, fields?.maxReasoningTokens ?? fields?.thinkingBudget))?.toUpperCase();
	if (thinkingLevel === "MINIMAL") includeThoughts = false;
	if (model.startsWith("gemini-3-pro")) {
		if (thinkingLevel === "MINIMAL") thinkingLevel = "LOW";
		else if (thinkingLevel === "MEDIUM") thinkingLevel = "HIGH";
	} else if (model.startsWith("gemini-3.1-pro")) {
		if (thinkingLevel === "MINIMAL") thinkingLevel = "LOW";
	}
	const modelVersion = getModelVersion(model);
	const modelSpecialty = getModelSpecialty(model);
	if (modelVersion === "2.5" && modelSpecialty === "image" || typeof thinkingBudget === "undefined" || typeof thinkingLevel === "undefined") return;
	const thinkingConfig = { includeThoughts };
	if (modelSpecialty === "") if (model.startsWith("gemini-2.5")) thinkingConfig.thinkingBudget = thinkingBudget;
	else thinkingConfig.thinkingLevel = thinkingLevel;
	return thinkingConfig;
}
/**
* Builds the `speechConfig` object for the Gemini `generationConfig`,
* normalising the user-friendly simplified forms into the wire format.
*
* @param fields The combined user/call-option fields.
* @returns A populated {@link Gemini.SpeechConfig}, or `{}`
*          when no speech config was provided.
*/
function convertFieldsToSpeechConfig(fields) {
	const config = fields.speechConfig;
	if (typeof config === "undefined") return;
	function isSpeechConfig(config) {
		return typeof config === "object" && (Object.hasOwn(config, "voiceConfig") || Object.hasOwn(config, "multiSpeakerVoiceConfig"));
	}
	function hasLanguage(config) {
		return typeof config === "object" && Object.hasOwn(config, "languageCode");
	}
	function hasVoice(config) {
		return Object.hasOwn(config, "voice");
	}
	if (isSpeechConfig(config)) return config;
	let languageCode;
	let voice;
	if (hasLanguage(config)) {
		languageCode = config.languageCode;
		voice = hasVoice(config) ? config.voice : config.voices;
	} else {
		languageCode = void 0;
		voice = config;
	}
	let ret;
	if (typeof voice === "string") ret = { voiceConfig: { prebuiltVoiceConfig: { voiceName: voice } } };
	else ret = { multiSpeakerVoiceConfig: { speakerVoiceConfigs: (Array.isArray(voice) ? voice : [voice]).map((v) => ({
		speaker: v.speaker,
		voiceConfig: { prebuiltVoiceConfig: { voiceName: v.name } }
	})) } };
	if (languageCode) ret.languageCode = languageCode;
	return ret;
}
/**
* Builds an object that can be included as part of the request
* which will add a `speechConfig` attribute if appropriate.
*/
function convertFieldsToServiceTier(platform, fields) {
	if (platform === "gcp") return {};
	const serviceTier = fields.serviceTier;
	if (!serviceTier || !settableServiceTier.includes(serviceTier)) return {};
	return { serviceTier };
}
//#endregion
export { convertFieldsToServiceTier, convertFieldsToSpeechConfig, convertFieldsToThinkingConfig, convertParamsToPlatformType };
