// @langchain/google/dist/converters/tools.cjs
const require_errors = require("../utils/errors.cjs");
let _langchain_core_utils_json_schema = require("@langchain/core/utils/json_schema");
let _langchain_core_utils_types = require("@langchain/core/utils/types");
let _langchain_core_utils_function_calling = require("@langchain/core/utils/function_calling");
let _langchain_core_language_models_base = require("@langchain/core/language_models/base");
let _langchain_core_utils_standard_schema = require("@langchain/core/utils/standard_schema");
//#region src/converters/tools.ts
/**
* Adjusts a JSON Schema object type to be compatible with Gemini's function schema format.
*
* Gemini's function schema format has specific requirements:
* - Type must be a string, not an array
* - Nullable types should use the `nullable` property instead of including "null" in a type array
* - Union types are not supported
*
* @param obj - The JSON Schema object to adjust
* @returns The adjusted object with type converted to Gemini-compatible format
* @throws {Error} If the type is a union type or null-only type, which Gemini cannot handle
*
* @internal
*/
function adjustObjectType(obj) {
	if (!Array.isArray(obj.type)) return obj;
	const len = obj.type.length;
	const nullIndex = obj.type.indexOf("null");
	if (len === 2 && nullIndex >= 0) {
		const typeIndex = nullIndex === 0 ? 1 : 0;
		obj.type = obj.type[typeIndex];
		obj.nullable = true;
	} else if (len === 1 && nullIndex === 0) throw new require_errors.InvalidInputError("Gemini does not support null-only types in function schemas. Provide a non-null type alongside nullable.");
	else if (len === 1) obj.type = obj?.type[0];
	else if (len === 0) throw new require_errors.InvalidInputError("Gemini does not support empty type arrays in function schemas. Provide at least one type.");
	else throw new require_errors.InvalidInputError("Gemini does not support union types in function schemas. Use a single type instead.");
	return obj;
}
/**
* Recursively removes unsupported properties from a JSON Schema object to make it
* compatible with Gemini's function schema format.
*
* Gemini's function schema format does not support:
* - `additionalProperties` property
* - Array types (must be converted to string types with nullable flag)
* - Union types
*
* @param obj - The JSON Schema object to clean
* @returns A cleaned GeminiFunctionSchema object
*
* @internal
*/
function removeAdditionalProperties(obj) {
	if (typeof obj === "object" && obj !== null) {
		const newObj = { ...obj };
		if ("additionalProperties" in newObj) delete newObj.additionalProperties;
		adjustObjectType(newObj);
		for (const key in newObj) if (key in newObj) {
			if (Array.isArray(newObj[key])) newObj[key] = newObj[key].map(removeAdditionalProperties);
			else if (typeof newObj[key] === "object" && newObj[key] !== null) newObj[key] = removeAdditionalProperties(newObj[key]);
		}
		return newObj;
	}
	return obj;
}
/**
* Converts a Zod schema or JSON Schema to Gemini function parameters format.
*
* This function handles the conversion of schema definitions from various formats
* (Zod schemas, JSON Schema 7) to Gemini's function schema format. It removes
* unsupported properties and adjusts type definitions to be compatible with Gemini.
*
* @param schema - The schema to convert (Zod schema or JSON Schema 7)
* @returns A schema object compatible with Gemini's API
*
* @internal
*/
function schemaToGeminiParameters(schema) {
	const { $schema, ...rest } = removeAdditionalProperties((0, _langchain_core_utils_types.isInteropZodSchema)(schema) || (0, _langchain_core_utils_standard_schema.isSerializableSchema)(schema) ? (0, _langchain_core_utils_json_schema.toJsonSchema)(schema) : schema);
	return rest;
}
/**
* Converts a JSON Schema object to Gemini function parameters format.
*
* @param schema - The JSON Schema object to convert
* @returns A schema object compatible with Gemini's API
*
* @internal
*/
function jsonSchemaToGeminiParameters(schema) {
	const { $schema, ...rest } = removeAdditionalProperties(schema);
	return rest;
}
/**
* Type guard to check if a tool is already in Gemini's native format.
*
* Gemini tools can be:
* - Function declarations (`functionDeclarations`)
* - Code execution tools (`codeExecution`)
* - Google Search retrieval tools (`googleSearchRetrieval` or `googleSearch`)
*
* The full list of tool types are those defined by the `Gemini.Tool` type.
* Updating that list requires a change to the `geminiToolAttributes` below.
*
* @param tool - The tool to check
* @returns `true` if the tool is already in Gemini format, `false` otherwise
*
* @internal
*/
function isGeminiTool(tool) {
	if (typeof tool !== "object" || tool === null) return false;
	return [
		"functionDeclarations",
		"codeExecution",
		"googleSearch",
		"urlContext",
		"googleMaps",
		"fileSearch",
		"computerUse",
		"mcpServers"
	].some((attr) => attr in tool);
}
/**
* Converts an array of LangChain tools to Gemini function declarations.
*
* This converter transforms various tool formats into Gemini's function declaration
* format. It supports:
* - LangChain structured tools (with Zod schemas)
* - OpenAI format tools (with JSON Schema)
* - Gemini-native function declarations (passed through)
* - Raw function declaration objects
*
* @remarks
* The converter handles each tool type differently:
* - **LangChain tools**: Extracts name, description, and schema, converting the schema
*   to Gemini-compatible format
* - **OpenAI tools**: Extracts function name, description, and parameters from the
*   OpenAI tool format
* - **Gemini tools**: If the tool contains `functionDeclarations`, extracts and returns
*   them directly. Non-function Gemini tools (like `codeExecution`) are skipped.
* - **Raw objects**: Assumes the object is a function declaration and validates it has
*   a `name` property
*
* @example
* ```typescript
* import { StructuredTool } from "@langchain/core/tools";
* import { z } from "zod";
*
* const weatherTool = new StructuredTool({
*   name: "get_weather",
*   description: "Get the weather for a location",
*   schema: z.object({
*     location: z.string().describe("The city name"),
*     unit: z.enum(["celsius", "fahrenheit"]).optional()
*   }),
*   func: async (args) => { /* ... *\/ }
* });
*
* const declarations = convertToolsToGeminiFunctionDeclarations([weatherTool]);
* // Returns: [{
* //   name: "get_weather",
* //   description: "Get the weather for a location",
* //   parameters: {
* //     type: "object",
* //     properties: {
* //       location: { type: "string", description: "The city name" },
* //       unit: { type: "string", enum: ["celsius", "fahrenheit"], nullable: true }
* //     },
* //     required: ["location"]
* //   }
* // }]
* ```
*
* @param tools - Array of tools in various formats (LangChain, OpenAI, or Gemini)
* @returns Array of Gemini function declarations ready for API submission
* @throws {Error} If a tool cannot be converted (invalid format)
*/
const convertToolsToGeminiFunctionDeclarations = (tools) => {
	const functionDeclarations = [];
	for (const tool of tools) {
		if (isGeminiTool(tool) && !("functionDeclarations" in tool)) continue;
		if (isGeminiTool(tool) && "functionDeclarations" in tool && Array.isArray(tool.functionDeclarations)) {
			functionDeclarations.push(...tool.functionDeclarations);
			continue;
		}
		if ((0, _langchain_core_utils_function_calling.isLangChainTool)(tool)) {
			const jsonSchema = schemaToGeminiParameters(tool.schema);
			functionDeclarations.push({
				name: tool.name,
				description: tool.description ?? `A function available to call.`,
				parameters: jsonSchema
			});
			continue;
		}
		if ((0, _langchain_core_language_models_base.isOpenAITool)(tool)) {
			functionDeclarations.push({
				name: tool.function.name,
				description: tool.function.description ?? `A function available to call.`,
				parameters: jsonSchemaToGeminiParameters(tool.function.parameters)
			});
			continue;
		}
		if (typeof tool === "object" && tool !== null && "name" in tool && typeof tool.name === "string") {
			const funcDecl = tool;
			functionDeclarations.push({
				name: funcDecl.name,
				description: funcDecl.description ?? `A function available to call.`,
				parameters: funcDecl.parameters
			});
			continue;
		}
		throw new require_errors.InvalidToolError(tool);
	}
	return functionDeclarations;
};
/**
* Converts an array of tools to Gemini's Tools array format.
*
* This converter groups function declarations into a single tool object while
* preserving Gemini-native tools (like code execution and Google Search retrieval)
* as separate tool objects. This matches Gemini's API structure where:
* - Function declarations are grouped together in a single tool
* - Non-function tools (codeExecution, googleSearchRetrieval) are separate tools
*
* @remarks
* The conversion process:
* 1. Separates Gemini-native tools from function declarations
* 2. Converts LangChain/OpenAI tools to function declarations
* 3. Groups all function declarations into a single Tool object
* 4. Returns an array with the grouped function declarations tool plus any
*    non-function Gemini tools
*
* @example
* ```typescript
* import { StructuredTool } from "@langchain/core/tools";
* import { z } from "zod";
*
* const weatherTool = new StructuredTool({
*   name: "get_weather",
*   description: "Get the weather",
*   schema: z.object({ location: z.string() }),
*   func: async () => {}
* });
*
* const geminiCodeTool: Gemini.Tool = {
*   codeExecution: { enabled: true }
* };
*
* const tools = convertToolsToGeminiTools([weatherTool, geminiCodeTool]);
* // Returns: [
* //   { functionDeclarations: [{ name: "get_weather", ... }] },
* //   { codeExecution: { enabled: true } }
* // ]
* ```
*
* @param tools - Array of tools in various formats (LangChain, OpenAI, or Gemini)
* @returns Array of Tool objects ready for API submission
*/
const convertToolsToGeminiTools = (tools) => {
	if (tools.length === 0) return [];
	const geminiTools = [];
	const functionDeclarations = [];
	for (const tool of tools) if (isGeminiTool(tool)) if ("functionDeclarations" in tool) geminiTools.push(tool);
	else geminiTools.push(tool);
	else if ("googleSearchRetrieval" in tool) geminiTools.push({ googleSearch: {} });
	else {
		const funcDecls = convertToolsToGeminiFunctionDeclarations([tool]);
		functionDeclarations.push(...funcDecls);
	}
	if (functionDeclarations.length > 0) geminiTools.push({ functionDeclarations });
	return geminiTools;
};
/**
* Converts tool choice to Gemini's function calling configuration.
*
* This converter transforms LangChain's tool choice format into Gemini's
* `toolConfig.functionCallingConfig` format. The configuration is only
* created when tools are present (`hasTools` is true).
*
* @remarks
* The conversion logic:
* - If `hasTools` is `false`, returns `undefined` (no config needed)
* - If `toolChoice` is provided, converts it to a Gemini mode:
*   - `"auto"` → `"AUTO"` - Model decides whether to call functions
*   - `"any"` or `"required"` → `"ANY"` - Model must call at least one function
*   - `"none"` → `"NONE"` - Model cannot call functions
*   - Function name string → `"ANY"` - Forces function use
*   - Object with `mode` → Maps the mode string to Gemini format
*   - Object with `function` → `"ANY"` - Forces function use
* - If `toolChoice` is not provided but tools exist, defaults to `"AUTO"`
*
* Gemini's function calling modes:
* - **AUTO**: The model can choose whether to call functions based on the conversation
* - **ANY**: The model must call at least one function before responding
* - **NONE**: The model cannot call any functions
*
* When a specific function is requested (via function name string or `{ function: { name } }`),
* the converter uses `"ANY"` mode to force function usage, as Gemini doesn't support
* specifying individual functions in the mode.
*
* @example
* ```typescript
* // With explicit tool choice
* convertToolChoiceToGeminiConfig("auto", true);
* // Returns: { functionCallingConfig: { mode: "AUTO" } }
*
* // Required mode - must call a function
* convertToolChoiceToGeminiConfig("any", true);
* // Returns: { functionCallingConfig: { mode: "ANY" } }
*
* // Without tool choice (defaults to AUTO when tools exist)
* convertToolChoiceToGeminiConfig(undefined, true);
* // Returns: { functionCallingConfig: { mode: "AUTO" } }
*
* // No tools - returns undefined
* convertToolChoiceToGeminiConfig("auto", false);
* // Returns: undefined
* ```
*
* @param toolChoice - The tool choice option from LangChain (string, object, or undefined)
* @param hasTools - Whether tools are present. If false, returns undefined
* @returns The Gemini tool configuration object, or undefined if no tools are present
*/
function convertToolChoiceToGeminiConfig(toolChoice, hasTools) {
	if (!hasTools || toolChoice === void 0) return;
	let mode;
	let allowedFunctionNames;
	let toolChoiceMode;
	let toolChoiceFunction;
	if (typeof toolChoice === "object") {
		toolChoiceMode = toolChoice.mode;
		toolChoiceFunction = toolChoice.function?.name;
	} else toolChoiceMode = toolChoice;
	if (toolChoiceMode === "auto") mode = "AUTO";
	else if (toolChoiceMode === "any" || toolChoiceMode === "required") mode = "ANY";
	else if (toolChoiceMode === "none") mode = "NONE";
	else if (typeof toolChoiceMode === "string") {
		mode = "ANY";
		toolChoiceFunction = toolChoiceMode;
	}
	if (toolChoiceFunction) if (Array.isArray(toolChoiceFunction)) allowedFunctionNames = toolChoiceFunction;
	else allowedFunctionNames = [toolChoiceFunction];
	const functionCallingConfig = {};
	if (allowedFunctionNames?.length) functionCallingConfig.allowedFunctionNames = allowedFunctionNames;
	if (mode) functionCallingConfig.mode = mode;
	return { functionCallingConfig };
}
//#endregion
exports.convertToolChoiceToGeminiConfig = convertToolChoiceToGeminiConfig;
exports.convertToolsToGeminiTools = convertToolsToGeminiTools;
exports.schemaToGeminiParameters = schemaToGeminiParameters;
